File: examples/agent-hooks/bin/dent8-native-memory-guard.py
#!/usr/bin/env python3
"""Small hook helper for dent8's agent integration examples.

The script is intentionally conservative: it never writes native memory files.
It can:

- run `dent8 verify` at session start;
- warn or block direct writes to agent-native memory/rules files;
- run `dent8 verify` after a write that touched those files.

Set DENT8_HOOK_MODE to one of:

- session-start
- guard-native-memory-write
- post-write-audit

Set DENT8_HOOK_ENFORCE=1 to turn a guard hit into exit code 2. Most agent hook
systems treat exit code 2 as a blocking policy failure.
"""

from __future__ import annotations

import json
import os
import re
import subprocess
import sys
from collections.abc import Iterable
from pathlib import Path
from typing import Any


DEFAULT_PATTERNS = (
    r"(^|/)AGENTS\.md$",
    r"(^|/)CLAUDE(\.local)?\.md$",
    r"(^|/)MEMORY\.md$",
    r"(^|/)GEMINI\.md$",
    r"(^|/)\.cursor/rules/.*\.(md|mdc)$",
    r"(^|/)\.devin/rules/.*\.md$",
    r"(^|/)\.windsurf/rules/.*\.md$",
)

PATH_KEYS = {
    "absolute_path",
    "file",
    "filePath",
    "file_path",
    "new_path",
    "old_path",
    "path",
    "relative_path",
    "target_file",
}


def main() -> int:
    payload = read_payload()
    mode = os.environ.get("DENT8_HOOK_MODE", "guard-native-memory-write")

    if mode == "session-start":
        return run_verify("session start")

    touched = sorted(native_memory_paths(payload))
    if mode == "post-write-audit":
        if touched:
            return run_verify(f"native memory/rules changed: {', '.join(touched)}")
        return 0

    if mode != "guard-native-memory-write":
        print(f"dent8 hook: unknown DENT8_HOOK_MODE={mode}", file=sys.stderr)
        return 2

    if not touched:
        return 0

    message = (
        "dent8 native memory/rules guard: direct writes to "
        f"{', '.join(touched)} bypass the claim-event firewall. "
        "Use dent8 MCP tools or an explicit reviewed export from dent8. "
        "Set DENT8_ALLOW_NATIVE_MEMORY_WRITE=1 to bypass this local guard."
    )
    print(message, file=sys.stderr)

    if os.environ.get("DENT8_ALLOW_NATIVE_MEMORY_WRITE") == "1":
        return 0
    if os.environ.get("DENT8_HOOK_ENFORCE") == "1":
        return 2
    return 0


def read_payload() -> Any:
    raw = sys.stdin.read()
    if not raw.strip():
        return {}
    try:
        return json.loads(raw)
    except json.JSONDecodeError as exc:
        print(f"dent8 hook: could not parse hook JSON: {exc}", file=sys.stderr)
        return {}


def native_memory_paths(payload: Any) -> set[str]:
    patterns = compiled_patterns()
    paths = set()
    for value in candidate_strings(payload):
        normalized = value.replace("\\", "/")
        if any(pattern.search(normalized) for pattern in patterns):
            paths.add(normalized)
    return paths


def compiled_patterns() -> list[re.Pattern[str]]:
    raw = os.environ.get("DENT8_NATIVE_MEMORY_PATTERNS")
    patterns = raw.split(os.pathsep) if raw else list(DEFAULT_PATTERNS)
    return [re.compile(pattern) for pattern in patterns if pattern]


def candidate_strings(value: Any) -> Iterable[str]:
    if isinstance(value, dict):
        for key, child in value.items():
            if key in PATH_KEYS and isinstance(child, str):
                yield child
            yield from candidate_strings(child)
    elif isinstance(value, list):
        for child in value:
            yield from candidate_strings(child)
    elif isinstance(value, str) and looks_like_path(value):
        yield value


def looks_like_path(value: str) -> bool:
    markers = (
        "/",
        "\\",
        "AGENTS.md",
        "CLAUDE.md",
        "CLAUDE.local.md",
        "GEMINI.md",
        "MEMORY.md",
        ".cursor/rules",
        ".devin/rules",
        ".windsurf/rules",
    )
    return any(marker in value for marker in markers)


def run_verify(reason: str) -> int:
    dent8 = os.environ.get("DENT8_BIN", "dent8")
    print(f"dent8 hook: verify ({reason})", file=sys.stderr)
    try:
        completed = subprocess.run(
            [dent8, "verify"],
            check=False,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
        )
    except FileNotFoundError:
        print(f"dent8 hook: {dent8!r} not found; skipping verify", file=sys.stderr)
        return 0

    if completed.stdout:
        print(completed.stdout.rstrip(), file=sys.stderr)
    if completed.stderr:
        print(completed.stderr.rstrip(), file=sys.stderr)
    return completed.returncode


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/agent-hooks/bin/dent8-native-memory-guard.ts
#!/usr/bin/env node
/*
 * Small TypeScript hook helper for dent8's agent integration examples.
 *
 * This is behavior-compatible with dent8-native-memory-guard.py. It is useful
 * for agent setups that already standardize on Node/TypeScript tooling.
 *
 * Set DENT8_HOOK_MODE to one of:
 *
 * - session-start
 * - guard-native-memory-write
 * - post-write-audit
 *
 * Set DENT8_HOOK_ENFORCE=1 to turn a guard hit into exit code 2. Most agent
 * hook systems treat exit code 2 as a blocking policy failure.
 */

declare const require: (specifier: string) => any;
declare const process: {
  env: Record<string, string | undefined>;
  exitCode?: number;
};

const childProcess = require("node:child_process");
const fs = require("node:fs");
const path = require("node:path");

const DEFAULT_PATTERNS: readonly string[] = [
  String.raw`(^|/)AGENTS\.md$`,
  String.raw`(^|/)CLAUDE(\.local)?\.md$`,
  String.raw`(^|/)MEMORY\.md$`,
  String.raw`(^|/)GEMINI\.md$`,
  String.raw`(^|/)\.cursor/rules/.*\.(md|mdc)$`,
  String.raw`(^|/)\.devin/rules/.*\.md$`,
  String.raw`(^|/)\.windsurf/rules/.*\.md$`,
];

const PATH_KEYS: ReadonlySet<string> = new Set([
  "absolute_path",
  "file",
  "filePath",
  "file_path",
  "new_path",
  "old_path",
  "path",
  "relative_path",
  "target_file",
]);

function main(): number {
  const payload = readPayload();
  const mode = process.env.DENT8_HOOK_MODE ?? "guard-native-memory-write";

  if (mode === "session-start") {
    return runVerify("session start");
  }

  const touched = Array.from(nativeMemoryPaths(payload)).sort();
  if (mode === "post-write-audit") {
    if (touched.length > 0) {
      return runVerify(`native memory/rules changed: ${touched.join(", ")}`);
    }
    return 0;
  }

  if (mode !== "guard-native-memory-write") {
    console.error(`dent8 hook: unknown DENT8_HOOK_MODE=${mode}`);
    return 2;
  }

  if (touched.length === 0) {
    return 0;
  }

  console.error(
    "dent8 native memory/rules guard: direct writes to " +
      `${touched.join(", ")} bypass the claim-event firewall. ` +
      "Use dent8 MCP tools or an explicit reviewed export from dent8. " +
      "Set DENT8_ALLOW_NATIVE_MEMORY_WRITE=1 to bypass this local guard.",
  );

  if (process.env.DENT8_ALLOW_NATIVE_MEMORY_WRITE === "1") {
    return 0;
  }
  if (process.env.DENT8_HOOK_ENFORCE === "1") {
    return 2;
  }
  return 0;
}

function readPayload(): unknown {
  const raw = fs.readFileSync(0, "utf8");
  if (raw.trim() === "") {
    return {};
  }

  try {
    return JSON.parse(raw);
  } catch (error: unknown) {
    const message =
      error instanceof Error ? error.message : "unknown JSON parse error";
    console.error(`dent8 hook: could not parse hook JSON: ${message}`);
    return {};
  }
}

function nativeMemoryPaths(payload: unknown): Set<string> {
  const patterns = compiledPatterns();
  const paths = new Set<string>();

  for (const value of candidateStrings(payload)) {
    const normalized = value.replaceAll("\\", "/");
    if (patterns.some((pattern) => pattern.test(normalized))) {
      paths.add(normalized);
    }
  }

  return paths;
}

function compiledPatterns(): RegExp[] {
  const raw = process.env.DENT8_NATIVE_MEMORY_PATTERNS;
  const patterns = raw ? raw.split(path.delimiter) : Array.from(DEFAULT_PATTERNS);
  return patterns.filter((pattern: string) => pattern !== "").map(
    (pattern: string) => new RegExp(pattern),
  );
}

function* candidateStrings(value: unknown): Iterable<string> {
  if (isRecord(value)) {
    for (const [key, child] of Object.entries(value)) {
      if (PATH_KEYS.has(key) && typeof child === "string") {
        yield child;
      }
      yield* candidateStrings(child);
    }
    return;
  }

  if (Array.isArray(value)) {
    for (const child of value) {
      yield* candidateStrings(child);
    }
    return;
  }

  if (typeof value === "string" && looksLikePath(value)) {
    yield value;
  }
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function looksLikePath(value: string): boolean {
  const markers = [
    "/",
    "\\",
    "AGENTS.md",
    "CLAUDE.md",
    "CLAUDE.local.md",
    "GEMINI.md",
    "MEMORY.md",
    ".cursor/rules",
    ".devin/rules",
    ".windsurf/rules",
  ];
  return markers.some((marker) => value.includes(marker));
}

function runVerify(reason: string): number {
  const dent8 = process.env.DENT8_BIN ?? "dent8";
  console.error(`dent8 hook: verify (${reason})`);

  const completed = childProcess.spawnSync(dent8, ["verify"], {
    encoding: "utf8",
  });
  const error = completed.error as { code?: string; message?: string } | undefined;

  if (error?.code === "ENOENT") {
    console.error(`dent8 hook: '${dent8}' not found; skipping verify`);
    return 0;
  }
  if (error) {
    console.error(`dent8 hook: verify failed to start: ${error.message}`);
    return 2;
  }

  if (completed.stdout) {
    console.error(String(completed.stdout).trimEnd());
  }
  if (completed.stderr) {
    console.error(String(completed.stderr).trimEnd());
  }

  return typeof completed.status === "number" ? completed.status : 2;
}

process.exitCode = main();
